### ISSUE TITLE
`VecElement` doesn't work with 1-tuples

### ISSUE THREAD

**Issue by eschnett** [JuliaLang/julia#16287]
@ArchRobison I find that the `VecElement` mechanism doesn't work with length-1 tuples; these are still represented as LLVM arrays in function arguments. Is this intended? If not, do you know where I could change this?

This is inconvenient because it requires many special cases in <https://github.com/eschnett/SIMD.jl>.

**Comment by ArchRobison** [comment]
I wasn't sure what to do for the singleton case, so I left it out.  But now we have a motivating case, so let's fix it.

The place to fix is in `alloc.c`, here:
```
unsigned jl_special_vector_alignment(size_t nfields, jl_value_t *t) {
    if (!is_vecelement_type(t))
        return 0;
    if (nfields>16 || (1<<nfields & 0x1157C) == 0)
        // Number of fields is not 2, 3, 4, 5, 6, 8, 10, 12, or 16.
        return 0;
```
Just change the C to an E (4 sharps :-) and update the comment.

- Arch

**Comment by eschnett** [comment]
I was going to ask about `0x1157C` anyway: Why is this set of numbers special? Is this just a choice? Is there a connection to LLVM's interpretation of the x86-64 ABI? Is there a connection to types and tuples used during Julia's bootstrap?

**Comment by ArchRobison** [comment]
The set has binary numerals in [2..16] with no more than two bits equal to one.  Those seemed to be the vector lengths that worked correctly with an unpatched LLVM according to your comments a while back, though in retrospect I should have used [1..16] as the starting point.

**Comment by eschnett** [comment]
This set also depends on the LLVM version. If it's based on LLVM bugs, then I'll add a respective comment to the source code as well, and I'll see whether LLVM 3.7 or 3.8 fare better here.

**Comment by ArchRobison** [comment]
Thanks for looking into this.  I've been using my spare time to work on #16138 .

**Comment by eschnett** [comment]
For the record: for both LLVM 3.7 and 3.8, vector sizes with N>4 and N%4==3 cause problems -- either LLVM crashes, or it generates wrong code.

**Comment by eschnett** [comment]
I quick test in C suggests that this LLVM crash occurs only for SIMD vectors that are passed by reference. Passing them by value avoids the error. Passing them by value is also likely to be more efficient.

This code demonstrates this:
```C
typedef long vlong7 __attribute__((__ext_vector_type__(7)));

vlong7 vadd7(vlong7 x, vlong7 y) { return x + y; }

vlong7 pvadd7(const vlong7 *restrict x, const vlong7 *restrict y) {
  return *x + *y;
}
```
When built on an Intel CPU with AVX-2 instructions with the `-march=native` flag, it triggers the LLVM crash. The first routine (`vadd7`) is fine, only the second (`pvadd7`) is problematic.

**Comment by vchuravy** [comment]
It seems that llvm only really support vectors that a powers of two: https://llvm.org/bugs/show_bug.cgi?id=11023 and the comment is still in the code https://github.com/llvm-mirror/llvm/blob/4366cdb0c28846d3308198acb39d7df7b7dacd6f/lib/CodeGen/TargetLoweringBase.cpp#L1125-L1126

But I would think that llvm shouldn't crash (and I just confirmed with llvm-svn), so we should open an upstream bug about this.

**Comment by eschnett** [comment]
LLVM should neither crash nor should silently generate wrong code.

SIMD vectors of length 3 are quite important; they are e.g. part of the OpenCL standard, and are commonly used in computer graphics.

Last time I checked LLVM for this bug, I found a routine that handles SIMD vectors of odd sizes. Vectors are either extended up to the length of the next available machine SIMD vector size, or are broken into several SIMD vectors. I don't recall where this routine was or how it was called. The logic in this routine was somewhat convoluted, and I couldn't confirm it works correctly.

See <https://llvm.org/bugs/show_bug.cgi?id=27708>.
